rubbish/src/Command/Parser.hs

```diff
@@ -1,94 +1,54 @@
-          , puSoftwareVersion :: SoftwareVersion
-          , puUpdates         :: [ProposeUpdateSystem]
-          }
-    | Help
-    | ListAddresses
-    | DelegateLight !Int !PublicKey !EpochIndex !(Maybe EpochIndex) -- first and last epoch of psk ttl
-    | DelegateHeavy !Int !PublicKey !EpochIndex -- last argument is current epoch
-    | AddKeyFromPool !Int
-    | AddKeyFromFile !FilePath
-    | AddrDistr !PublicKey !AddrStakeDistribution
-    | Quit
-    deriving Show
-
-data ProposeUpdateSystem = ProposeUpdateSystem
-    { pusSystemTag     :: SystemTag
-    , pusInstallerPath :: Maybe FilePath
-    , pusBinDiffPath   :: Maybe FilePath
-    } deriving Show
+import           Universum
+
+import           Data.ByteString.Base58 (bitcoinAlphabet, decodeBase58)
+import qualified Data.List.NonEmpty     as NE
+import qualified Data.Map               as Map
+import qualified Data.Text              as Text
+import           Prelude                (read)
+import           Serokell.Util.Parse    (parseIntegralSafe)
+import           Text.Parsec            (many1, parse, parserFail, try, (<?>))
+import           Text.Parsec.Char       (alphaNum, anyChar, digit, noneOf, space, spaces,
+                                         string)
+import           Text.Parsec.Combinator (eof, manyTill)
+import           Text.Parsec.Text       (Parser)
+
+import           Pos.Binary             (decodeFull)
+import           Pos.Client.CLI         (stakeholderIdParser)
+import           Pos.Core.Version       (parseBlockVersion, parseSoftwareVersion)
+import           Pos.Crypto             (AbstractHash, HashAlgorithm, PublicKey,
+                                         decodeAbstractHash)
+import           Pos.Txp                (TxOut (..))
+import           Pos.Types              (AddrStakeDistribution (..), Address, Coin,
+                                         CoinPortion, decodeTextAddress, mkCoin,
+                                         mkMultiKeyDistr, unsafeCoinPortionFromDouble)
+import           Pos.Update             (SystemTag, mkSystemTag)
+import           Pos.Util.Util          (eitherToFail)
+
+import           Command.Types          (Command (..), ProposeUpdateParams (..),
+                                         ProposeUpdateSystem (..), SendMode (..),
+                                         SendToAllGenesisParams (..))
+
 
 lexeme :: Parser a -> Parser a
 lexeme p = spaces *> p >>= \x -> spaces $> x
 
 text :: String -> Parser Text
 text = lexeme . fmap toText . string
 
-many1Till :: Parser a -> Parser b -> Parser [a]
-many1Till p end = (:) <$> p <*> manyTill p end
+-- many1Till :: Parser a -> Parser b -> Parser [a]
+-- many1Till p end = (:) <$> p <*> manyTill p end
 
 anyText :: Parser Text
 anyText = lexeme $ fmap toText $ manyTill anyChar (void (try space) <|> try eof)
 
 filePath :: Parser FilePath
-filePath = many1Till (alphaNum <|> oneOf "_.") (void (try space) <|> try eof)
+filePath = lexeme (many1 anyChar)
```

Comment:
Mda, now this thing will consume everything up to `eof`.

Reply:
The proper way to implement it is to add escaping for whitespace in filenames.

Reply:
I am wondering why nobody has written it yet 😒